apollo-federation/src/query_plan/conditions.rs

```diff
@@ -282,12 +280,13 @@ pub(crate) fn remove_unneeded_top_level_fragment_directives(
                         let final_selection =
                             inline_fragment.with_updated_selection_set(updated_selections);
                         selection_map.insert(Selection::InlineFragment(Arc::new(final_selection)));
+                    } else {
+                        // We can skip some of the fragment directives directive.
+                        let final_selection = inline_fragment
+                            .with_updated_directives(DirectiveList::from_iter(needed_directives))
+                            .with_updated_selection_set(updated_selections);
+                        selection_map.insert(Selection::InlineFragment(Arc::new(final_selection)));
```

Comment:
This is unrelated to the problem that is being solved, but it is a bug. The JS code returned inside the `if` statement. Also, the selection set was not being updated like it should have been.